python/ray/_raylet.pyx

```diff
@@ -1516,6 +1523,169 @@ cdef class EmptyProfileEvent:
     def __exit__(self, *args):
         pass
 
+class RetryGcsConnectionError(Exception):
+    "If this error is raised, we should try to reconnect to GCS."
+    def __init__(self, exception):
+        self.exception = exception
+
+def _auto_reconnect(f):
+    @wraps(f)
+    def wrapper(self, *args, **kwargs):
+        if "TEST_RAY_COLLECT_KV_FREQUENCY" in os.environ:
+            ray._private.utils._CALLED_FREQ[f.__name__] += 1
+        remaining_retry = self._nums_reconnect_retry
+        while True:
+            try:
+                return f(self, *args, **kwargs)
+            except RetryGcsConnectionError as e:
+                if remaining_retry <= 0:
+                    raise e.exception
+                logger.debug(
+                    f"Failed to send request to gcs, reconnecting. Error {e.exception}"
+                )
+                try:
+                    self._connect()
+                except Exception:
+                    logger.error(f"Connecting to gcs failed. Error {e.exception}")
+                time.sleep(1)
+                remaining_retry -= 1
+                continue
+    return wrapper
+
+cdef class GcsClient:
```

Comment:
i wonder if we have unit test cover the GcsClient, especially around the failure handling.

Reply:
We have Python (integration) tests that exercise the re-connection logic to the GCS (I know because I fixed the failures for this PR), let me try to find them again and post here.

Reply:
cc @iycheng for ^

Reply:
But yeah I think we now rely on GCS client to detect GCS failures...